x-pack/plugins/elastic_assistant/server/lib/langchain/execute_custom_llm_chain/index.ts

```diff
@@ -88,21 +107,96 @@ export const callAgentExecutor = async ({
-    agentType: 'chat-conversational-react-description',
-    memory,
-    verbose: false,
-  });
+  // isStream check is not on agentType alone because typescript doesn't like
+  const executor = isStream
+    ? await initializeAgentExecutorWithOptions(tools, llm, {
+        agentType: 'openai-functions',
+        memory,
+        verbose: true,
+      })
+    : await initializeAgentExecutorWithOptions(tools, llm, {
+        agentType: 'chat-conversational-react-description',
+        memory,
+        verbose: false,
+      });
 
   // Sets up tracer for tracing executions to APM. See x-pack/plugins/elastic_assistant/server/lib/langchain/tracers/README.mdx
   // If LangSmith env vars are set, executions will be traced there as well. See https://docs.smith.langchain.com/tracing
   const apmTracer = new APMTracer({ projectName: traceOptions?.projectName ?? 'default' }, logger);
 
   let traceData;
+  if (isStream) {
+    const {
+      end: streamEnd,
+      push,
+      responseWithHeaders,
+    } = streamFactory<{ type: string; payload: string }>(request.headers, logger, false, false);
+
+    let didEnd = false;
+
+    const handleStreamEnd = (finalResponse: string) => {
+      if (onLlmResponse) {
+        onLlmResponse(finalResponse);
+      }
+      // @yuliia this would be a good place for pushing the response to the chat history
+      streamEnd();
+      didEnd = true;
+    };
+
+    let message = '';
+
+    executor
+      .invoke(
+        {
+          input: latestMessage[0].content,
+          chat_history: [],
+          signal: abortSignal,
+        },
+        {
+          callbacks: [
+            {
+              handleLLMNewToken(payload) {
+                if (payload.length && !didEnd) {
+                  push({ payload, type: 'content' });
+                  // store message in case of error
+                  message += payload;
+                }
+              },
+              handleChainEnd(llmResult) {
+                handleStreamEnd(llmResult.output);
+              },
+            },
+            apmTracer,
+            ...(traceOptions?.tracers ?? []),
+          ],
+          runName: DEFAULT_AGENT_EXECUTOR_ID,
+          tags: traceOptions?.tags ?? [],
+        }
+      )
+      .catch((err) => {
+        // if I throw an error here, it crashes the server. Not sure how to get around that.
+        // If I put await on this function the error works properly, but when there is not an error
+        // it waits for the entire stream to complete before resolving
+        const error = transformError(err);
+
+        if (error.message === 'AbortError') {
+          // user aborted the stream, we must end it manually here
+          return handleStreamEnd(message);
+        }
+        logger.error(`Error streaming from LangChain: ${error.message}`);
+        push({ payload: error.message, type: 'content' });
+        handleStreamEnd(error.message);
+      });
+
+    // TODO figure out how to pass trace_data @spong
```

Comment:
@spong I'm not sure about `trace_data` for langchain streaming because `withAssistantSpan`/`withSpan` is async. You will notice nothing is awaited within this `isStream` block. await would block the stream from being returned to the client immediately. I'm not sure if there is a non-async apm function (I looked but did not see). Otherwise, maybe we should look at moving tracking to the `post_actions_execute` level (perhaps in a follow up PR)

If I do try passing the `invoke` call to `withSpan`, I get this type error:
```
TS2345: Argument of type Promise<void | ChainValues> is not assignable to parameter of type (span?: Span | undefined) => Promise<unknown>
Type Promise<void | ChainValues> provides no match for the signature (span?: Span | undefined): Promise<unknown>
```

Reply:
Okie, so got it working by just manually starting a `span` if streaming is enabled:

```ts
let streamingSpan: Span | undefined;
if (agent.isStarted()) {
  streamingSpan = agent.startSpan(`${DEFAULT_AGENT_EXECUTOR_ID} (Streaming)`) ?? undefined;
}
```

Then when streaming is complete you set the `outcome` and `end()` it:

```ts
if (!streamingSpan?.outcome || streamingSpan?.outcome === 'unknown') {
  streamingSpan.outcome = 'success';
}
streamingSpan?.end();
```

And then they start showing up in the APM UI (complete w/ tool usage 🙌):

<p align="center">
  <img width="700" src="https://github.com/elastic/kibana/assets/2946766/e59c40eb-6fbf-4709-a3d1-bacc5915178c" />
</p> 

@dgieselaar, is there anything else to consider when trying to capture around streaming content like this? Is there a corresponding synchronous APM util (like `withSpan`) that would be better to use here, or is creating a simple span like this fine as-is?

Reply:
Note: There are currently two bugs around working with `traceData`: 

1. Conditionally showing the 'View APM trace for this button' is broke as `traceData` is now always an empty object instead of `undefined`, so we need to throw an `&& Object.keys(message.traceData).length > 0` here: 
https://github.com/elastic/kibana/blob/15ca37fccc45ffdd67313a29b960a67cfc210cdc/x-pack/plugins/security_solution/public/assistant/comment_actions/index.tsx#L86-L89 

2. There's a type mis-match around persistence in `onLlmResponse` that is preventing non-streaming `traceData` from persisting. It is typed as `Message['traceData']`, which is camel case:

https://github.com/elastic/kibana/blob/5512e42fa2d3db512c9ba80c8232755cfa129635/x-pack/plugins/elastic_assistant/server/routes/post_actions_connector_execute.ts#L152-L155

https://github.com/elastic/kibana/blob/f625be7c3e61b4252640ed3794b927c5291a9df5/x-pack/packages/kbn-elastic-assistant-common/impl/schemas/conversations/common_attributes.gen.ts#L53-L62


but is being passed in as `ResponseBody`, which is snake case (see my other [comment](https://github.com/elastic/kibana/pull/174126/files#r1550928501) here):

https://github.com/elastic/kibana/blob/aca83aa6c7ad1fa25493f80d7d0fe47b4d195015/x-pack/plugins/elastic_assistant/server/lib/langchain/types.ts#L17-L20


https://github.com/elastic/kibana/blob/8005580879cbbc31513c5c54bc2913df94c1679e/x-pack/plugins/elastic_assistant/server/lib/langchain/execute_custom_llm_chain/index.ts#L204-L206


---

Let's see what @dgieselaar has to say about the above, but I'm happy to pair on the fix here if you'd like -- shouldn't be much more than the snippets above once we fix the response types.

Reply:
I've got Garrett's updates in this commit: https://github.com/elastic/kibana/pull/174126/commits/8f0e7759285c385bf2a4acce9cf930842334dc0f